src/Npgsql/NpgsqlConnector.cs

```diff
@@ -674,14 +674,23 @@ async Task RawOpen(NpgsqlTimeout timeout, bool async, CancellationToken cancella
                         break;
                     case 'S':
                         var clientCertificates = new X509Certificate2Collection();
-                        var certPath = Settings.ClientCertificate ?? PostgresEnvironment.SslCert;
-
-                        if (certPath is null && PostgresEnvironment.SslCertDefault is string certPathDefault)
-                            certPath = certPathDefault;
+                        var certPath = Settings.ClientCertificate ?? PostgresEnvironment.SslCert ?? PostgresEnvironment.SslCertDefault;
 
                         if (certPath != null)
                         {
-                            cert = new X509Certificate2(certPath, Settings.ClientCertificateKey ?? PostgresEnvironment.SslKey);
+                            var password = Settings.ClientCertificateKeyPassword;
+#if NET5_0
+                            if (Path.GetExtension(certPath).ToUpperInvariant() != ".PFX")
+                            {
+                                // It's PEM time
+                                var keyPath = Settings.ClientCertificateKey ?? PostgresEnvironment.SslKey;
+                                cert = string.IsNullOrEmpty(password)
+                                        ? X509Certificate2.CreateFromPemFile(certPath, keyPath)
+                                        : X509Certificate2.CreateFromEncryptedPemFile(certPath, password, keyPath);
+                            }
+#endif
+                            if (cert is null)
+                                cert = new X509Certificate2(certPath, password);
```

Comment:
I don't understand all this stuff very well, but doesn't this change before for non-PEM keys, i.e. use `ClientCertificateKeyPassword` instead of `Settings.ClientCertificateKey ?? PostgresEnvironment.SslKey`?

Reply:
`ClientCertificateKey` is technically a path to the key (and so is `PostgresEnvironment.SslKey`). Think about it as if there are 3 parts:
1. Certficate
2. A private key for the certificate
3. A password from the private key (this part is optional)

.PFX contains 1 and 2 part together, but for .PEM it might not be the same (the first and the second part might be in different files).

Reply:
Take a look at the documentation for [libpq](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PARAMKEYWORDS) (`sslcert`, `sslkey` and `sslpassword`).

Reply:
But doesn't this change behavior for PFX certificates which were supported before? In other words, will this PR change behavior (and possibly break) existing code using Npgsql today?

Reply:
Yeah, it is a breaking change (although, considering that's a connection string change, it should be fairly easy to fix on the user side).

The only alternatives I see here, is to either:
a) add a new connection string param for the key path (instead of the password), something like `ClientCertificateKeyPath` or
b) continue to use `ClientCertificateKey` as a password for the PFX

Reply:
I'll try to take a more thorough look later to understand everything.

Reply:
Sure, this can wait.

Reply:
I looked at this again... I guess I'm OK with the breaking change here - it makes it more consistent that ClientCertificateKeyPassword always contains a password, whereas ClientCertificateKey always contains a path to a key file. However, it seems like for PEM, ClientCertificateKeyPassword is the password to the key file, whereas for PFX it's the password to the certificate file - so the name ClientCertificateKeyPassword is maybe problematic... Am I seeing that right?

Also, we should probably make sure we're aligned with libpq on the environment variables - and ideally also with the connection string parameters, as much as possible...

Reply:
> However, it seems like for PEM, ClientCertificateKeyPassword is the password to the key file, whereas for PFX it's the password to the certificate file - so the name ClientCertificateKeyPassword is maybe problematic... Am I seeing that right?

Yeah, you're right - libpq uses `sslcert`, `sslkey` and `sslpassword`, which are a little bit more 'general' in their naming. Are you proposing to rename them?

Reply:
Hmm, we should consider it... So in libpq `sslkey` is always a path, and `sslpassword` is always a password (either to the key file, or to the certificate)?

Aligning with libpq on *environment variables* seems very important to me - this way Npgsql applications can be dropped in a PG environment and behave in a consistent way. Connection string parameters are a bit less important, though they also need to be at least minimally consistent with the environment variables.

Maybe we should just rename these to their PostgreSQL equivalents, i.e. have `SSL Certificate` (corresponding to `PGSSLCERT`), `SSL Key` (corresponding to `PGSSLKEY`) and `SSL Password` (no env var). We could leave the older names (obsoleting the properties) for backwards compat. What do you think?